L1Trigger/L1TTrackMatch/interface/DisplacedVertexProducer.h

```diff
@@ -0,0 +1,264 @@
+  }
+  float y(float phi_T=0){
+    return ( charge * rho * TMath::Cos(phi - charge*phi_T) - (d0 + charge * rho) * TMath::Cos(phi));
+  }
+  float z(float phi_T=0){
+    float theta = 2 * TMath::ATan(TMath::Exp(-eta));
+    return (z0 + rho*phi_T/TMath::Tan(theta));
+  }
+  float deltaPhi_T(Double_t phi1, Double_t phi2)
+  {
+    Double_t dPhi = phi1 - phi2;
+    if (dPhi >= TMath::Pi())
+      dPhi -= 2. * TMath::Pi();
+    if (dPhi <= -TMath::Pi())
+      dPhi += 2. * TMath::Pi();
+    return dPhi;
+  }
+  float phi_T(float x, float y){
+    float num = x - (d0 + charge * rho) * TMath::Sin(phi);
+    float den = y + (d0 + charge * rho) * TMath::Cos(phi);
+    return ((phi-TMath::ATan2(num,-den))/charge);
+  }
+  float z(float x, float y){
+    float t = std::sinh(eta);
+    float r = TMath::Sqrt(pow(x,2)+pow(y,2));
+    return (z0+(t*r*(1+(pow(d0,2)/pow(r,2))+(1.0/6.0)*pow(r/(2*rho),2)))); // can do higher order terms if necessary from displaced math
+  }
+  Track_Parameters(float pt_in, float d0_in, float z0_in, float eta_in, float phi_in, int pdgid_in, float vx_in, float vy_in, float vz_in, float charge_in=0, int index_in=-1, Track_Parameters* tp_in=nullptr, int nstubs_in=0, float chi2rphi_in=0, float chi2rz_in=0, float bendchi2_in=0, float MVA1_in=0, float MVA2_in=0)
+  {
+    pt = pt_in;
+    d0 = d0_in;
+    z0 = z0_in;
+    eta = eta_in;
+    phi = phi_in;
+    if(charge_in > 0){
+      charge = 1;
+    }
+    else if (charge_in < 0){
+      charge = -1;
+    }
+    else{
+      charge = 0;
+    }
+    index = index_in;
+    pdgid = pdgid_in;
+    vx = vx_in;
+    vy = vy_in;
+    vz = vz_in;
+    tp = tp_in;
+    rho = fabs(1/charge_in);
+    x0 = (rho+charge*d0)*TMath::Cos(phi-(charge*TMath::Pi()/2));
+    y0 = (rho+charge*d0)*TMath::Sin(phi-(charge*TMath::Pi()/2));
+    nstubs = nstubs_in;
+    chi2rphi = chi2rphi_in;
+    chi2rz = chi2rz_in;
+    bendchi2 = bendchi2_in;
+    MVA1 = MVA1_in;
+    MVA2 = MVA2_in;
+  }
+  Track_Parameters(){};
+  ~Track_Parameters(){};
+};
+
+constexpr bool operator==(const Track_Parameters* lhs, const Track_Parameters& rhs)
+{
+  return (lhs->pt==rhs.pt && lhs->d0==rhs.d0 && lhs->z0==rhs.z0 && lhs->eta==rhs.eta && lhs->phi==rhs.phi);
+}
+constexpr bool operator==(const Track_Parameters& lhs, const Track_Parameters* rhs)
+{
+  return (lhs.pt==rhs->pt && lhs.d0==rhs->d0 && lhs.z0==rhs->z0 && lhs.eta==rhs->eta && lhs.phi==rhs->phi);
+}
+constexpr bool operator==(const Track_Parameters& lhs, const Track_Parameters& rhs)
+{
+  return (lhs.pt==rhs.pt && lhs.d0==rhs.d0 && lhs.z0==rhs.z0 && lhs.eta==rhs.eta && lhs.phi==rhs.phi);
+}
+
+std::valarray<float> calcPVec(Track_Parameters a, double_t v_x, double_t v_y)
+{
+  std::valarray<float> r_vec = {float(v_x)-a.x0,float(v_y)-a.y0};
+  std::valarray<float> p_vec = {-r_vec[1],r_vec[0]};
+  if(a.charge>0){
+    p_vec *= -1;
+  }
+  p_vec /= TMath::Sqrt(pow(p_vec[0],2)+pow(p_vec[1],2));
+  p_vec *= a.pt;
+  return p_vec;
+}
+
+class Vertex_Parameters
+{
+public:
```

Comment:
My personal preference would be to get rid of this class as well and use these variables directly in the DisplacedVertexProducer class to avoid another vertex-like class, but I also see why you may want to set it up like this. It doesn't quite follow how the other GTT object classes are set up so I would suggest maybe changing this class to a struct instead as it's really only holding variables to be used by DisplacedVertexProducer.

There is no vertex class that has all of these variables.

Reply:
can this (class => functions) be done as a "code improvement" in a second stage? L1 is requesting that this PR is merged in order for @ryanm124 to show the plots at an upcoming conference. that request is to have code to be able to reproduce the results, and it seems code improvements wouldn't be necessary for that?

Reply:
This is ok with me for this comment. There does seem to be some potential logic issues brought up in other comments that might need more addressing. And at minimum, the code checks and formatting should be done and It doesn't seem like the formatting has been applied. You should run these two commands:
```
scram build code-checks
scram build code-format
```

Reply:
Code formatting has been applied [93ea36d](https://github.com/cms-l1t-offline/cmssw/pull/1257/commits/93ea36dc899f03444ec82e37a5fcd126ec64cae6)